ndb/src/google/cloud/ndb/model.py

```diff
@@ -2302,17 +2302,135 @@ def _db_get_value(self, v, unused_p):
+    """
+
     __slots__ = ()
 
-    def __init__(self, *args, **kwargs):
-        raise NotImplementedError
+    def _validate(self, value):
+        """Validate a ``value`` before setting it.
+
+        Args:
+            value (~datetime.date): The value to check.
+
+        Raises:
+            .BadValueError: If ``value`` is not a :class:`~datetime.date`.
+        """
+        if not isinstance(value, datetime.date):
+            raise exceptions.BadValueError(
+                "Expected date, got {!r}".format(value)
+            )
+
+    def _to_base_type(self, value):
+        """Convert a value to the "base" value type for this property.
+
+        Args:
+            value (~datetime.date): The value to be converted.
+
+        Returns:
+            ~datetime.datetime: The converted value: a datetime object with the
+            time set to ``00:00``.
+
+        Raises:
+            TypeError: If ``value`` is not a :class:`~datetime.date`.
+        """
+        if not isinstance(value, datetime.date):
+            raise TypeError(
+                "Cannot convert to datetime expected date value; "
+                "received {}".format(value)
+            )
+        return datetime.datetime(value.year, value.month, value.day)
+
+    def _from_base_type(self, value):
+        """Convert a value from the "base" value type for this property.
+
+        Args:
+            value (~datetime.datetime): The value to be converted.
+
+        Returns:
+            ~datetime.date: The converted value: the date that ``value``
+            occurs on.
+        """
+        return value.date()
+
+    @staticmethod
+    def _now():
+        """datetime.datetime: Return current date."""
+        return datetime.datetime.utcnow().date()
 
 
 class TimeProperty(DateTimeProperty):
+    """A property that contains :class:`~datetime.time` values.
+
+    .. automethod:: _to_base_type
+    .. automethod:: _from_base_type
+    .. automethod:: _validate
+    """
+
     __slots__ = ()
 
-    def __init__(self, *args, **kwargs):
-        raise NotImplementedError
+    def _validate(self, value):
+        """Validate a ``value`` before setting it.
+
+        Args:
+            value (~datetime.time): The value to check.
+
+        Raises:
+            .BadValueError: If ``value`` is not a :class:`~datetime.time`.
+        """
+        if not isinstance(value, datetime.time):
+            raise exceptions.BadValueError(
+                "Expected time, got {!r}".format(value)
+            )
+
+    def _to_base_type(self, value):
+        """Convert a value to the "base" value type for this property.
+
+        Args:
+            value (~datetime.time): The value to be converted.
+
+        Returns:
+            ~datetime.datetime: The converted value: a datetime object with the
+            date set to ``1970-01-01``.
+
+        Raises:
+            TypeError: If ``value`` is not a :class:`~datetime.time`.
+        """
+        if not isinstance(value, datetime.time):
```

Comment:
There [was][1] an `assert` here but it didn't make sense because

1. `_validate()` already checked the type
1. The `_time_to_datetime()` helper checks the type.

FWIW I moved the `_time_to_datetime()` implementation here wholesale

[1]: https://github.com/GoogleCloudPlatform/datastore-ndb-python/blob/cf4cab3f1f69cd04e1a9229871be466b53729f3f/ndb/model.py#L2203

Reply:
:+1: